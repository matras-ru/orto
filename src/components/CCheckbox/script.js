import { selfInstall } from '@/utils';
import commonAttributes from '@/mixins/commonAttributes.js';

import ThemeClass from '@/themes/default/CCheckbox';

const {
    disabledClass,
    baseIconClass,
    defaultIconClass,
    checkedIconClass,
    disabledIconClass
} = ThemeClass;

export default {
    name: 'Checkbox',

    install(Vue, theme) {
        selfInstall(Vue, theme, this);
    },

    props: {
        ...commonAttributes.props,
        label: {
            type: String,
            default: null
        },
        hint: {
            type: String,
            default: null
        },
        modelValue: {
            type: [Array, Boolean],
            default: null
        },
        value: {
            type: [String, Number],
            default: null
        },
        trueValue: {
            type: Boolean,
            default: true
        },
        falseValue: {
            type: Boolean,
            default: false
        }
    },

    model: {
        prop: 'modelValue',
        event: 'change'
    },

    data() {
        return {
            isChecked: false
        };
    },

    computed: {
        currentClass() {
            let classes = ['flex mb-1-4'];

            if (this.disabled) {
                classes.push(disabledClass);
            }

            return classes;
        },
        currentIconClass() {
            let classes = [baseIconClass];

            if (this.disabled) {
                classes.push(disabledIconClass);
            } else if (this.isChecked || this.shouldBeChecked) {
                classes.push(checkedIconClass);
            } else {
                classes.push(defaultIconClass);
            }

            return classes;
        },
        shouldBeChecked() {
            if (this.modelValue instanceof Array) {
                return this.modelValue.includes(this.value);
            }

            return this.modelValue === this.trueValue;
        }
    },

    methods: {
        onChange(e) {
            this.isChecked = e.target.checked;

            if (this.modelValue instanceof Array) {
                let newValue = [...this.modelValue];

                if (this.isChecked) {
                    newValue.push(this.value);
                } else {
                    newValue.splice(newValue.indexOf(this.value), 1);
                }

                this.$emit('change', newValue);
            } else {
                this.$emit('change', this.isChecked ? this.trueValue : this.falseValue);
            }
        }
    },

    render(h) {
        const attrs = {
            id: this.id,
            autofocus: this.autofocus,
            name: this.name,
            type: 'checkbox',
            label: this.label,
            value: this.value,
            disabled: this.disabled
        };

        const on = {
            change: this.onChange
        };

        const componentData = {
            class: 'absolute opacity-0 pointer-events-none',
            attrs: attrs,
            on
        };

        return h('label', { class: this.currentClass }, [
            h('input', componentData),
            h('div', { class: 'flex' }, [
                h('span', { class: this.currentIconClass }),
                h('span', this.label)
            ])
        ]);
    }
};
