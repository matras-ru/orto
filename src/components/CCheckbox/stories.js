import { storiesOf } from '@storybook/vue';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';

import CCheckbox from './';

const stories = storiesOf('CCheckbox', module);
stories.addDecorator(withKnobs);

stories
    .add('Checkbox knobs', () => ({
        components: { CCheckbox },
        props: {
            id: {
                default: text('id', 'checkbox1')
            },
            autofocus: {
                default: boolean('autofocus', false)
            },
            label: {
                default: text('label', 'Checkbox1')
            },
            name: {
                default: text('name', 'checkbox1')
            },
            disabled: {
                default: boolean('disabled', false)
            },
            value: {
                default: text('value', 'checkbox1')
            }
        },
        data() {
            return {
                checkboxModel: false
            };
        },
        template:
            '<CCheckbox @onChange="action" :id="id" :autofocus="autofocus" :name="name" :label="label" :disabled="disabled" :value="value" v-model="checkboxModel" />',
        methods: { action: action('change') }
    }))
    .add('Default checkbox', () => ({
        components: { CCheckbox },
        data() {
            return {
                checkboxModel: false
            };
        },
        template:
            '<CCheckbox @onChanget="action" id="checkbox2" label="Checkbox2" name="checkbox2" value="checkbox2" v-model="checkboxModel" />',
        methods: { action: action('change') }
    }))
    .add('Disabled checkbox', () => ({
        components: { CCheckbox },
        template:
            '<CCheckbox @onChange="action" id="checkbox3" label="Checkbox3" name="checkbox3" value="checkbox3" disabled />',
        methods: { action: action('change') }
    }));
