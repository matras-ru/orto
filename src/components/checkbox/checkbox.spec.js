/* eslint-env jest */
import { mount, createLocalVue } from '@vue/test-utils';
import CCheckbox from './checkbox';

describe('Checkbox', () => {
    it('is rendered', () => {
        const wrapper = mount(CCheckbox, {
            context: {
                props: {
                    id: 'checkbox1',
                    name: 'checkbox1',
                    value: 'checkbox1',
                    label: 'default'
                }
            }
        });

        const input = wrapper.find('input');
        const label = wrapper.find('label');

        expect(input.attributes('id')).toBeDefined();
        expect(input.attributes('id')).toBe('checkbox1');
        expect(input.attributes('type')).toBeDefined();
        expect(input.attributes('type')).toBe('checkbox');
        expect(input.attributes('name')).toBeDefined();
        expect(input.attributes('name')).toBe('checkbox1');
        expect(input.attributes('name')).toBe('checkbox1');
        expect(label.attributes('for')).toBe('checkbox1');

        expect(wrapper.text()).toBe('default');

        expect(wrapper.classes().sort()).toEqual('flex flex-wrap mb-0-4'.split(' ').sort());
        expect(label.classes().sort()).toEqual('relative pl-1-4 cursor-pointer'.split(' ').sort());
        expect(input.classes().sort()).toEqual(
            'form-checkbox absolute top-0-2 left-0 w-0-8 h-0-8'.split(' ').sort()
        );

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('is disabled', () => {
        const wrapper = mount(CCheckbox, {
            context: {
                props: {
                    id: 'checkbox1',
                    name: 'checkbox1',
                    value: 'checkbox1',
                    disabled: true
                }
            }
        });

        const input = wrapper.find('input');
        const label = wrapper.find('label');

        expect(input.attributes('disabled')).toBeDefined();
        expect(input.attributes('disabled')).toBe('disabled');
        expect(label.classes()).toContain('cursor-not-allowed');
        expect(label.classes()).toContain('opacity-50');

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('is error', () => {
        const wrapper = mount(CCheckbox, {
            context: {
                props: {
                    id: 'checkbox1',
                    name: 'checkbox1',
                    value: 'checkbox1',
                    error: true
                }
            }
        });

        const input = wrapper.find('input');
        expect(input.classes()).toContain('form-checkbox-is-error');

        expect(wrapper.html()).toMatchSnapshot();

        wrapper.destroy();
    });

    it('v-model - single value - true/false', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: false
                };
            },

            render(h) {
                return h(CCheckbox, {
                    props: {
                        id: 'checkbox1',
                        name: 'checkbox1',
                        value: 'checkbox1',
                        modelValue: this.model
                    },
                    on: {
                        change: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input');

        expect(wrapper.vm.model).toBe(false);

        input.setChecked();
        expect(wrapper.vm.model).toBe(true);

        wrapper.destroy();
    });

    it('v-model - multiply values - array', () => {
        const localVue = new createLocalVue();

        const App = localVue.extend({
            data() {
                return {
                    model: []
                };
            },

            render(h) {
                return h(CCheckbox, {
                    props: {
                        id: 'checkbox1',
                        name: 'checkbox1',
                        value: 'checkbox1',
                        modelValue: this.model
                    },
                    on: {
                        change: val => {
                            this.model = val;
                        }
                    }
                });
            }
        });

        const wrapper = mount(App, {
            localVue: localVue
        });

        const input = wrapper.find('input');

        expect(wrapper.vm.model).toEqual([]);

        input.setChecked();
        expect(wrapper.vm.model).toEqual(['checkbox1']);

        wrapper.destroy();
    });
});
