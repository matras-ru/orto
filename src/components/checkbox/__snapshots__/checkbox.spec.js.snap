// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox is disabled 1`] = `<div class="flex flex-wrap mb-0-4"><label for="checkbox1" class="relative pl-1-4 cursor-not-allowed opacity-50"><input id="checkbox1" name="checkbox1" type="checkbox" disabled="disabled" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-checkbox" value="checkbox1"></label></div>`;

exports[`Checkbox is error 1`] = `<div class="flex flex-wrap mb-0-4"><label for="checkbox1" class="relative pl-1-4 cursor-pointer"><input id="checkbox1" name="checkbox1" type="checkbox" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-checkbox form-checkbox-is-error" value="checkbox1"></label></div>`;

exports[`Checkbox is rendered 1`] = `<div class="flex flex-wrap mb-0-4"><label for="checkbox1" class="relative pl-1-4 cursor-pointer"><input id="checkbox1" name="checkbox1" type="checkbox" class="absolute top-0-2 left-0 w-0-8 h-0-8 form-checkbox" value="checkbox1">default</label></div>`;
